include/hydrogen/PoolAllocator.hpp

```diff
@@ -0,0 +1,1140 @@
+#endif
+  }
+  return error;
+}
+} // namespace gpudebug
+
+/**
+ * \brief Debug macro
+ */
+#ifndef gpuDebug
+#define gpuDebug(e) gpudebug::Debug((gpuError_t)(e), __FILE__, __LINE__)
+#endif
+
+/**
+ * Prints human-readable size (for reporting)
+ */
+static inline void HumanReadableSize(size_t bytes, std::ostream &os) {
+  const std::string sizes[] = {"B", "KiB", "MiB", "GiB", "TiB"};
+  int unit = 0;
+  float size = bytes;
+  while (size > 1024) {
+    size /= 1024;
+    ++unit;
+  }
+  auto oldprec = os.precision(unit > 0 ? 2 : 0);
+  auto oldf = os.setf(std::ios_base::fixed, std::ios_base::floatfield);
+  os << size << " " << sizes[unit];
+  os.precision(oldprec);
+  os.setf(oldf);
+}
+
+namespace hydrogen {
+
+/******************************************************************************
+ * PooledDeviceAllocator
+ ******************************************************************************/
+
+/**
+ * \brief A simple caching allocator for device memory allocations.
+ *
+ * \par Overview
+ * The allocator is thread-safe and stream-safe and is capable of managing
+ * cached device allocations on multiple devices.  It behaves as follows:
+ *
+ * \par
+ * - Allocations from the allocator are associated with an \p active_stream.
+ * Once freed, the allocation becomes available immediately for reuse within the
+ * \p active_stream with which it was associated with during allocation, and it
+ * becomes available for reuse within other streams when all prior work
+ * submitted to \p active_stream has completed.
+ * - Allocations are categorized and cached by bin size.  A new allocation
+ * request of a given size will only consider cached allocations within the
+ * corresponding bin.
+ * - (EXTENDED) Bin limits have a combined geometric/linear progression; or can
+ *   be given as a set of sizes. It behaves as follows:
+ *     - If a set of sizes is given in \p bin_sizes, they are used to construct
+ * the allocation bins. If an allocation is larger than the largest bin, the
+ * behavior matches the rest of the algorithm. Allocations in [0, bin_min] are
+ *       will allocate ``bin_min`` bytes.
+ *     - Bin limits progress geometrically in accordance with the (integer)
+ *       growth factor \p bin_growth provided during construction. Unused device
+ *       allocations within a larger bin cache are not reused for allocation
+ *       requests that categorize to smaller bin sizes.
+ *       Allocation requests below (\p bin_growth ^ \p min_bin) are rounded up
+ * to ( \p bin_growth ^ \p min_bin ).
+ *     - If two consecutive geometric bins exceed \p bin_mult_threshold and
+ *       \p bin_mult is given, a linear binning scheme is created where bins
+ *       follow \p bin_growth ^ some_bin + \p bin_mult * n
+ *     - Allocations above min( \p bin_growth ^ \p max_bin , \p
+ * max_bin_alloc_size ) are not rounded up to the nearest bin and are simply
+ * freed when they are deallocated instead of being returned to a bin-cache.
```

Comment:
```suggestion
 *     - Allocations above min( \p bin_growth ^ \p max_bin , \p
 *       max_bin_alloc_size ) are not rounded up to the nearest bin and are simply
 *       freed when they are deallocated instead of being returned to a bin-cache.
```